### PR TITLE
gbXML validation against XSD schema

### LINKED ISSUE #4441: GbXML ReverseTranslator needs validation and/or better error handling
Enhancement Request
--------------

See https://github.com/NREL/OpenStudio/pull/4440#discussion_r712822703

Really we need error handling in gbxml... There is currently no validation against the gbXML schema happening and there is very little checking that is done manually (like ensuring that required values/attributes are supplied, or checking that a YearSchedule references a weekScheduleIdRef you make sure it can be found).

## Detailed Description

I found this because a test I added was failing without any logging when I had a typo in a weekScheduleIdRef

```diff
  <Schedule type="Fraction" id="Simple_Schedule">
    <YearSchedule id="Simple_Schedule_YearSchedule">
      <BeginDate>2017-01-01</BeginDate>
      <EndDate>2017-12-31</EndDate>
-      <WeekScheduleId weekScheduleIdRef="SOMETHING_THAT_CANT_BE_FOUND" />
    </YearSchedule>
    [ .... ]
```

Another issue might be when you pass Schedule with zero YearSchedule children. Or if YearSchedule doesn't have a BeginDate children. etc etc.

https://gbxml.org/schema/6-01/GreenBuildingXML_Ver6.01.xsd#element_YearSchedule

## Possible Implementation

Schema validation against the gxbml XSD ([here](https://gbxml.org/schema/6-01/GreenBuildingXML_Ver6.01.xsd)would be an option. `pugixml` doesn't support that, so that would require using a different library for xml such as xeres-c (which is on conan-center-index: https://github.com/conan-io/conan-center-index/tree/master/recipes/xerces-c).

Another option is to manually enforce that required values and attributes are there, and handle errors appropriately (at least logging to inform the user).

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4441

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/lib/CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,7 @@ target_link_libraries(
  CONAN_PKG::cpprestsdk
  CONAN_PKG::openssl
  CONAN_PKG::pugixml
+ CONAN_PKG::xercesc
```
Comment (jmarrec) [inline src/lib/CMakeLists.txt:31]:
```suggestion
 CONAN_PKG::XercesC
```

Comment (jmarrec) [reply src/lib/CMakeLists.txt:31]: https://github.com/conan-io/conan-center-index/blob/2ebe7d8d6ecf4d53f59a411b4a4c47bae2f7abd1/recipes/xerces-c/all/conanfile.py#L149-L161

**Review by jmarrec — CHANGES_REQUESTED** [review]

File: src/utilities/xml/Validator.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include "../core/Logger.hpp"
+#include "../core/Optional.hpp"
+
+#include <boost/optional.hpp>
+
+#include <string>
+
+namespace openstudio {
+
+class UTILITIES_API Validator
```
Comment (jmarrec) [inline src/utilities/xml/Validator.hpp:45]: openstudio::Validator isn't very clear.

maybe openstudio::xml::Validator, or more simply openstudio::XmlValidator.

File: src/utilities/xml/Validator.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+Validator::Validator(const openstudio::path& xsdPath) : m_xsdPath(openstudio::filesystem::system_complete(xsdPath)) {
+  if (!openstudio::filesystem::exists(xsdPath)) {
+    LOG_AND_THROW("'" << toString(xsdPath) << "' does not exist");
+  } else if (!openstudio::filesystem::is_regular_file(xsdPath)) {
+    LOG_AND_THROW("'" << toString(xsdPath) << "' XSD cannot be opened");
+  }
+
+  xercesc::XMLPlatformUtils::Initialize();
+
+  // create the parser here?
```
Comment (jmarrec) [inline src/utilities/xml/Validator.cpp:75]: If we stick to a Ctor that takes the schema (and given there isn't a setXsdPath), then I would say yes, create the parser in the ctor.

Comment (jmarrec) [reply src/utilities/xml/Validator.cpp:75]: Additionally, make xercesc::XercesDOMParser parser; a member variable.

File: src/utilities/xml/Validator.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  if (errorCount > 0) {
+    LOG(Error, "'" << toString(xmlPath) << "' has " << errorCount << " error(s)");
+    return false;
+  }
+
+  return true;
+}
+
+bool Validator::validate(const std::string& xmlString) const {
+  return true;
```
Comment (jmarrec) [inline src/utilities/xml/Validator.cpp:140]: TODO

File: src/utilities/xml/Validator.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  explicit Validator(const openstudio::path& xsdPath);
+
+  explicit Validator(const std::string& xsdString);
+
+  //@}
+  /** @name Destructor */
+  //@{
+
+  /// Virtual destructor
+  virtual ~Validator() {}
```
Comment (jmarrec) [inline src/utilities/xml/Validator.hpp:61]:
```suggestion
```

File: src/utilities/xml/Validator.hpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  bool validate(const std::string& xmlString) const;
+
+  //@}
+  /** @name Operators */
+  //@{
+
+  //@}
+
+ private:
+  REGISTER_LOGGER("openstudio.xml.Validator");
```
Comment (jmarrec) [inline src/utilities/xml/Validator.hpp:90]: Adjust in coordination with the renaming I mentioned earlier (either add a namespace level or rename class to XmlValidator)

File: src/utilities/xml/Validator.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  if (parser.loadGrammar(toString(m_xsdPath).c_str(), xercesc::Grammar::SchemaGrammarType) == NULL) {
+    LOG(Error, "'" << toString(m_xsdPath) << "' XSD cannot be loaded");
+    return false;
+  }
+
+  // Enable schema processing.
+  parser.setDoSchema(true);
+  parser.setDoNamespaces(true);
+  parser.setValidationConstraintFatal(true);
+  parser.setValidationSchemaFullChecking(true);
```
Comment (jmarrec) [inline src/utilities/xml/Validator.cpp:119]: Do we need `parser.setValidationScheme(XercesDOMParser::Val_Always);` ?

https://xerces.apache.org/xerces-c/apiDocs-3/classAbstractDOMParser.html#a35600d583c42ceea9b04ee10814238a6

https://xerces.apache.org/xerces-c/apiDocs-3/classAbstractDOMParser.html#a76513b711425077116569013c93a943c

**Review by jmarrec — CHANGES_REQUESTED** [review]

File: src/utilities/xml/Validator.hpp (hunk 5 above)
Comment (joseph-robertson) [reply src/utilities/xml/Validator.hpp:61]: Did you mean `~XMLValidator()` here?

Comment (jmarrec) [reply src/utilities/xml/Validator.hpp]: I meant there is no reason to define a destructor, and even less a virtual one. This should follow the rule of 0/1. If you were to define a destructor, then you should follow the rule of 5 and defined the copy/assignment operators for lvalue and rvalue.

**Pull request closed without merging**